Userland/Libraries/LibProtocol/WebSocketClient.h

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (c) 2021, The SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/HashMap.h>
+#include <LibIPC/ServerConnection.h>
+#include <WebSocket/WebSocketClientEndpoint.h>
+#include <WebSocket/WebSocketServerEndpoint.h>
+
+namespace Protocol {
+
+class WebSocket;
+
+class WebSocketClient
+    : public IPC::ServerConnection<WebSocketClientEndpoint, WebSocketServerEndpoint>
+    , public WebSocketClientEndpoint {
+    C_OBJECT(WebSocketClient);
+
+public:
+    virtual void handshake() override;
+
+    RefPtr<WebSocket> connect(const URL& url, const String& origin = {}, const Vector<String>& protocols = {}, const Vector<String>& extensions = {}, const HashMap<String, String>& request_headers = {});
+
+    u32 ready_state(Badge<WebSocket>, WebSocket&);
+    void send(Badge<WebSocket>, WebSocket&, ByteBuffer, bool is_text);
+    void close(Badge<WebSocket>, WebSocket&, u16 code, String reason);
+    bool set_certificate(Badge<WebSocket>, WebSocket&, String, String);
+
+private:
+    WebSocketClient();
+
+    virtual void handle(const Messages::WebSocketClient::Connected&) override;
+    virtual void handle(const Messages::WebSocketClient::Received&) override;
+    virtual void handle(const Messages::WebSocketClient::Errored&) override;
+    virtual void handle(const Messages::WebSocketClient::Closed&) override;
+    virtual void handle(const Messages::WebSocketClient::CertificateRequested&) override;
+
+    HashMap<i32, RefPtr<WebSocket>> m_connections;
```

Comment:
Does this need to store null values? Otherwise, can we make it a `HashMap<i32, NonnullRefPtr<WebSocket>>`?

Reply:
Right now, using a `RefPtr<WebSocket>` allows to do things like `if ((connection = m_connections.get(message.connection_id()).value_or(nullptr)))` without a bunch of extra casting. Otherwise, it errors out (rightfully) because `value_or(nullptr)` doesn't work on a `Optional<NonnullRefPtr<T>>`. This is a simplfied version of the code in RequestClient :)

I'll try to rework that code to use `has_value()` and `value()`, though, because storing it as a `NonnullRefPtr` and checking the optional makes more sense :)

Reply:
Ok, the proposed code for getting the values is now :
```C++
    auto maybe_connection = m_connections.get(message.connection_id());
    if (maybe_connection.has_value())
        maybe_connection.value()->did_open({});
 ```
 Which is way clearer :)

Reply:
Cool beans! Note that `Optional` actually has `operator->` as well :)

Reply:
Since it's an `Optional<NonnullRefPtr<T>>`, it seems like the arrow operator returns the `NonnullRefPtr<T>` directly which doesn't redirect to the underlying call. The same code with `maybe_connection->did_open()` (skipping the intermediate `.value()`) fails to compile.